### PR TITLE
lwt-unix: use safe close when creating channels from fds

### PR DESCRIPTION
**Pull request by vbmithr**
Without this patch, closing both channels resulting from the same file descriptor will always raise an exception for the second close.

### CODE REVIEW

**Comment by hcarty** [comment]
I haven't tested this yet, but good catch!  I'll try to test tomorrow as I have tons of log lines warning about these `EBADF` errors.

**Comment by dinosaure** [comment]
Hmmhmm, I understand the purpose of this PR but I don't like to silently catch `EBADF` and return unit. I would like, at least, a warning to continue to keep tracking this kind of error. Otherwise, it seems that `conduit` has a bad design for this purpose, but it's an other problem.

**Comment by vbmithr** [comment]
I need a concrete solution to avoid logging thousands of useful lines in my logs, as well as issuing thousands of bogus syscalls.

File: lwt-unix/conduit_lwt_unix.ml [hunk 1]
```diff
@@ -160,6 +160,13 @@ let init ?src ?(tls_server_key=`None) () =
     | {ai_addr;_}::_ -> Lwt.return { src=Some ai_addr; tls_server_key }
     | [] -> Lwt.fail_with "Invalid conduit source address specified"
 
+let safe_close fd =
+  Lwt.catch
+    (fun () -> Lwt_unix.close fd)
```
Comment (hcarty) [inline lwt-unix/conduit_lwt_unix.ml]: Can we check with [Lwt_unix.state](http://ocsigen.org/lwt/4.1.0/api/Lwt_unix#VALstate) here before the call the `Lwt_unix.close`?  Then we can avoid the double-close without having to ignore the `EBADF`.

Something like (typed directly into github without testing):
```ocaml
let close_if_open fd =
  match Lwt_unix.state with
  | Lwt_unix.Aborted _
  | Lwt_unix.Opened -> Lwt_unix.close fd
  | Lwt_unix.Closed -> Lwt.return_unit
```

**Comment by hcarty** [comment]
@dinosaure Would https://github.com/mirage/ocaml-conduit/pull/292#r282625786 be acceptable to avoid the double-close without swallowing all possible `EBADF` cases?  If so, please feel free to use it in this PR @vbmithr

**Comment by dinosaure** [comment]
Yes I prefer this idea where we ensure that `BADF` came from it. Then, a notice on documentation will be perfect :+1: !

**Comment by avsm** [comment]
I’ve been digging into Lwt_io and it really seems like a regression, since channels are supposed to be lazily double closed without issue. Not sure what the root cause of this is

**Comment by hcarty** [comment]
@avsm In this case, though, it's two channels which share an underlying descriptor.  Would that guarantee still hold?

**Comment by avsm** [comment]
Ah! It's not safe to have two channels share the same fd then, I suspect.  The problem with catching the EBADF and ignoring it is that there is a potential race that is even worse:

- Channel A and B share fd 1
- Channel A closes fd 1
- New fd is created that reuses 1 
- Channel B closed fd 1 -- which is the new fd, and no EBADF is raised but a random other fd is closed

Given that Lwt_io channels do the close at most once, we seem to have to ensure that the channels themselves dont share fds to prevent double closes like the above

**Comment by vbmithr** [comment]
This is a bug in Lwt's API then. Should open a bug report there.

**Comment by hcarty** [comment]
I had a chat with @avsm and @dinosaure.  The issue is that we have two channels backed by the same file descriptor without handling the double-`close` appropriately.  It could be considered a bug in the Lwt API but I'm not sure it's an implementation bug.  May be worth clarifying the Lwt documentation.

I have a branch using `Lwt_io.open_connection`, based on @avsm's suggestion.  I'll push a PR once I've done a bit more testing.  It seems to address the double-close issue.

**Comment by avsm** [comment]
@vbmithr could you test #294 to see if it causes any other regressions in your setup?

**Comment by vbmithr** [comment]
Ok sure, will do.

**Comment by avsm** [comment]
Will be fixed via #294, thanks for the report!

**Comment by vbmithr** [comment]
Working for me, thanks! Closing this, superseded by #294

**Pull request closed without merging**